aptos-move/block-executor/src/code_cache_global_manager.rs

```diff
@@ -0,0 +1,530 @@
+    fn value_from_ready(&self) -> Option<T> {
+        match self {
+            State::Ready(v) => Some(v.clone()),
+            State::Clean | State::Executing(_) | State::Done(_) => None,
+        }
+    }
+
+    /// If the state is [State::Executing], returns its value. Otherwise, returns [None].
+    fn value_from_executing(&self) -> Option<T> {
+        match self {
+            State::Executing(v) => Some(v.clone()),
+            State::Clean | State::Ready(_) | State::Done(_) => None,
+        }
+    }
+
+    /// Sets the current state to [State::Ready].
+    fn set_ready(&mut self, value: T) {
+        *self = Self::Ready(value);
+    }
+
+    /// Sets the current state to [State::Executing].
+    fn set_executing(&mut self, value: T) {
+        *self = Self::Executing(value);
+    }
+
+    /// Sets the current state to [State::Done].
+    fn set_done(&mut self, value: T) {
+        *self = Self::Done(value);
+    }
+}
+
+/// Manages module caches and the execution environment, possible across multiple blocks.
+pub struct ModuleCacheManager<T, K, DC, VC, E> {
+    /// The state of global caches.
+    state: Mutex<State<T>>,
+
+    /// During concurrent executions, this module cache is read-only. However, it can be mutated
+    /// when it is known that there are no concurrent accesses. [ModuleCacheManager] must ensure
+    /// the safety.
+    module_cache: Arc<GlobalModuleCache<K, DC, VC, E>>,
+    /// The execution environment, initially set to [None]. The environment, as long as it does not
+    /// change, can be kept for multiple block executions.
+    environment: ExplicitSyncWrapper<Option<AptosEnvironment>>,
+}
+
+impl<T, K, DC, VC, E> ModuleCacheManager<T, K, DC, VC, E>
+where
+    T: Clone + Debug + Eq,
+    K: Hash + Eq + Clone,
+    VC: Deref<Target = Arc<DC>>,
+    E: WithSize,
+{
+    /// Returns a new instance of [ModuleCacheManager].
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            state: Mutex::new(State::Clean),
+            module_cache: Arc::new(GlobalModuleCache::empty()),
+            environment: ExplicitSyncWrapper::new(None),
+        }
+    }
+
+    /// If state is [State::Clean], or [State::Ready] with matching previous value, sets the state
+    /// to [State::Ready] with the current value and returns true. Otherwise, raises an alert and
+    /// returns false.
+    pub fn mark_ready(&self, previous: &T, current: T) -> bool {
+        let mut state = self.state.lock();
+
+        if state.is_clean() || state.is_done_with_value(previous) {
+            state.set_ready(current);
+            return true;
+        }
+
+        if state.is_done() {
+            // If the state is done, but the values to not match, we still set the state as ready, but
+            // also flush global caches because they execute not on top of the previous state.
+            self.module_cache.flush_unchecked();
```

Comment:
at this point I feel like module cache should have & to the environment to make sure flushes are coordinated (maybe other way too).

Reply:
Indeed, so what I am doing is ensuring that we have a single manager which holds both. Otherwise, on their own, module caches can exist even without environment?